src/lib/hash/sha3/sha3_round.h

```diff
@@ -0,0 +1,123 @@
+/*
+* (C) 2010,2016,2021 Jack Lloyd
+*
+* Botan is released under the Simplified BSD License (see license.txt)
+*/
+
+#include <botan/types.h>
+#include <botan/rotate.h>
+
+/*
+A bug in Clang 12 and XCode 13 cause a miscompilation of SHA3_round
+with optimization levels -O2 and higher.
+
+For details
+ https://github.com/randombit/botan/issues/2802 and
+ https://bugs.llvm.org/show_bug.cgi?id=51957
+*/
+
+#if    defined(__clang__) && \
+    (( defined(__apple_build_version__) && __clang_major__ == 13) || \
+     (!defined(__apple_build_version__) && __clang_major__ == 12))
+  #define BOTAN_HAS_BROKEN_CLANG_SHA3
+#endif
+
+#if defined(BOTAN_HAS_BROKEN_CLANG_SHA3)
+  #include <tuple>
+#endif
+
+namespace Botan {
+
+#if defined(BOTAN_HAS_BROKEN_CLANG_SHA3)
+
+namespace {
+
+__attribute__((noinline)) decltype(auto) xor_CNs(const uint64_t A[25])
+   {
+   return std::tuple<uint64_t, uint64_t, uint64_t, uint64_t, uint64_t>(
```

Comment:
For the record: For this to [work with older GCC](https://github.com/randombit/botan/pull/2849/files) the `std::tuple<...>` type needed to be repeated both for the return statement and the function declaration. Though, this code path being compiled only with the broken clang version, it shouldn't be a problem.

Reply:
One reason I wanted to have totally distinct codepaths for broken-Clang vs everything else - we only have to worry about the workaround in the specific case that we're being compiled under this particular compiler+version instead of having to work everywhere.

Oh that reminds me though, I think GCC 4.4 is still missing from the new GH actions CI and if that continues we'll certainly have more build regressions...